Test that bluechictl list-units aligns with systemctl output

mwperina opened the pull request:
Tests that bluechictl list-units return all units for all nodes and this
list matches list of units gathered from each node using systemctl.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/780
Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Test that bluechictl list-units aligns with systemctl output

Tests that bluechictl list-units return all units for all nodes and this
list matches list of units gathered from each node using systemctl.

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on tests/bluechi_test/bluechictl.py:
```diff
@@ -33,14 +33,22 @@ def _run(self, log_txt: str, cmd: str, check_result: bool, expected_result: int)
+                "list-units",
+                check_result,
+                expected_result
+            )
```
I'd remove the `else` here:
```suggestion
        return self._run(
            "Listing units on all nodes",
            "list-units",
            check_result,
            expected_result
        )
```

mwperina replied:
Done

engelmi commented on tests/bluechi_test/systemctl.py:
```diff
@@ -27,6 +27,15 @@ def _do_operation_on_unit(self, unit_name: str, operation: str, check_result: bo
+    def _do_operation(self, operation: str, check_result: bool, expected_result: int) \
+            -> Tuple[Optional[int], Union[Iterator[bytes], Any, Tuple[bytes, bytes]]]:
+
+        result, output = self.client.exec_run(f"{SystemCtl.binary_name} {operation}")
```
Could we wrap this in debug logs similar to `_on_unit`?

mwperina replied:
Hmm, I forgot about this. Debug logging is already included in exec_run: https://github.com/eclipse-bluechi/bluechi/blob/main/tests/bluechi_test/client.py#L82

So I'd rather post another commit to remove debug logging from _do_operation_on_unit to prevent repeating info in logs. WDYT?

engelmi replied:
Oh, good point! Yes, it would be better to remove it from `_do_operation_on_unit` - but only the second one (https://github.com/eclipse-bluechi/bluechi/blob/main/tests/bluechi_test/systemctl.py#L24). The first one is still useful, I think, and no duplication - plus it gets logged always (provided the debug log level).

mwperina replied:
But current logging in exec_run is always printed in debug, `container.exec_run` is not raising exceptions,
so now for each `exec_run` invocation we have following:

```
2024-02-26 14:42:30+0100,847 DEBUG   [bluechi_test.client] Executed command 'systemctl start bluechi-controller' with result '0' and output 'b''' (client:82)
```

So is there a value to have following 2 lines instead of 1?

```
2024-02-26 14:42:30+0100,656 DEBUG   [bluechi_test.systemctl] start unit 'bluechi-controller' (systemctl:22)                                                                                                                    
2024-02-26 14:42:30+0100,847 DEBUG   [bluechi_test.client] Executed command 'systemctl start bluechi-controller' with result '0' and output 'b''' (client:82)
```

engelmi replied:
Hmm... you are right. The debug log in systemctl could also be removed.

engelmi commented on tests/tests/tier0/bluechi-list-units-on-all-nodes/test_bluechi_list_units_on_all_nodes.py:
```diff
@@ -0,0 +1,110 @@
+        line = esc_seq.sub('', line)
+
+        match = line_pat.match(line)
+        if match:
```
Could we use an "early exit" here and get rid of the `else`? I think thats a bit easier to read and removes nesting. For example:
```python
if not match:
    raise Exception(f"Error parsing systemctl list-units output, invalid line: '{line}'")

# ...
```

mwperina replied:
Done

engelmi commented on tests/tests/tier0/bluechi-list-units-on-all-nodes/test_bluechi_list_units_on_all_nodes.py:
```diff
@@ -0,0 +1,110 @@
+            # Ignore header lines
+            continue
+        match = line_pat.match(line)
+        if match:
```
Could we use an "early exit" here and get rid of the `else`? I think thats a bit easier to read and removes nesting. For example:
```python
if not match:
    raise Exception(f"Error parsing bluechictl list-units output, invalid line: '{line}'")

# ...
```

mwperina replied:
Done

engelmi commented on tests/tests/tier0/bluechi-list-units-on-all-nodes/test_bluechi_list_units_on_all_nodes.py:
```diff
@@ -0,0 +1,110 @@
+            elif match.group("unit_name") in node_units:
+                raise Exception(f"Error parsing bluechictl list-units output, unit already reported, line: '{line}'")
+            else:
+                node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))
```
I think this could be simplified a bit, e.g.:

```suggestion
            if node_units is not None and match.group("unit_name") in node_units:
                raise Exception(f"Error parsing bluechictl list-units output, unit already reported, line: '{line}'")
            
            if not node_units:
                node_units = {}
                result[match.group("node_name")] = node_units

            node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))
```

mwperina replied:
You cannot use `in` operator for variable with `None` value (which happens for the 1st unit parsed for a node), that's why I have `if not not_units` condition first

engelmi replied:
Hmm... how about explicitly adding the `node_units is not None` check? (updated the snippet accordingly) Otherwise the order of those two ifs can still be reversed - just to avoid having the duplicate `node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))`

mwperina replied:
Done

mwperina committed:
Remove unnecessary logging in _do_operation_on_unit

Method exec_run in ContainerClient logs executed command, execution
result and output, so it not necessary to log those information again in
systemctl._do_operation_on_unit().

Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi reviewed (approved):
LGTM
